src/main/java/org/logstash/tcp/InputLoop.java

```diff
@@ -0,0 +1,198 @@
+package org.logstash.tcp;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.StandardSocketOptions;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public final class InputLoop implements Runnable, Closeable {
+
+    /**
+     * Read buffer size.
+     */
+    private static final int BUFFER_SIZE = 8096;
```

Comment:
@jakelandis then we only read 8k, no exceptions thrown or anything if that's your worry? :)